posts/2021/05/rethinking-jupyter-documentation.md

```diff
@@ -0,0 +1,192 @@
+it is typically reached via `identifer?` or `help(identifier)`
+
+Compared to rendered documentation, the help documentation is:
+
+ - Hard to read,
+ - Has no navigation
+ - RST Directives have not been interpreted.
+ - No inline graph, no rendered math.
+
+
+There is also no access to non-docstring based documentation, **no narrative**,
+**no tutorials**, **no image gallery or examples**, no search, no syntax
+highlighting, no way to interact or modify documentation to test effects of
+parameters.
+
+# Limitation for authors
+
+Due to Jupyter and IPython limitations to display documentation I believe
+authors are often contained to document functions.
+
+Syntax in docstrings is often kept simple for readability, this first version is
+often preferred:
+
+```rst
+You can use ``np.einsum('i->', a)`` ...
+```
+
+If the longer form, which makes the reference into a link when viewing rendered
+documentation, it is difficult to read when shown as help documentation:
+
+```rst
+You can use :py:func:`np.einsum('i->', a) <numpy.einsum>` ...
+```
+
+This also lead to long discussions about which syntax to use in advance area,
+like formulas in [Sympy's docstrings](https://github.com/sympy/sympy/issues/14964).
+
+Many projects have to implement dynamic docstrings; for example to include all
+the parameter a function or class, would pass down using ``**kwargs``, (search
+matplotlib source for `_kwdoc` , or pandas DataFrame for example).
+
+This can make it relatively difficult for authors and contributors to properly
+maintain and provide comprehensive docs.
+
+I'm not sure I can completely predict all the side effects this has on how library
+maintainers write docs; but I believe there is also a strong opportunity for a
+tools to help there. See for example [vélin](https://github.com/Carreau/velin)
+which attempts to auto reformat and fix common NumyDoc's format mistakes and
+typos – but that's a subject of a future post.
+
+# Stuck between a Rock and a Hard place
+
+While Sphinx and related project are great at offering hosted HTML
+documentation; extensive usage of those make interactive documentation harder to
+consume;
+
+While it is possible to [run Sphinx on the fly when rendering
+docstrings](https://github.com/spyder-ide/docrepr), most Sphinx features
+only work when building a full project, with the proper configuration and
+extension and can be computationally intensive. This make running Sphinx locally
+impractical.
+
+Hosted website often may not reflect locally installed version of the
+libraries and requires careful linking, deprecation and narrative around
+platform or version specific features.
+
+# This is fixable
+
+For the past few month I've been working on rewriting how IPython (and hence
+Jupyter) can display documentation. It works both in terminal (IPython) and
+browser context (notebook, JupyterLab, Spyder) with proper rendering, and currently
+understand most directives; it could be customized to understand any new ones:
```

Comment:
```suggestion
For the past few months I've been working on rewriting how IPython (and hence
Jupyter) can display documentation. It works both in terminal (IPython) and
browser context (notebook, JupyterLab, Spyder) with proper rendering, and currently
understands most directives; it could be customized to understand any new ones:
```